Doc: Update documentation how to make commit message (#12046)
OpenTTD/OpenTTD: CODINGSTYLE.md

The first line of a message must match:
```
<keyword>( #<issue>|<commit>(, (#<issue>|<commit>))*)?: ([<section])? <Details>
```

Keywords are split up in two sections.

Player-facing changes:
* Feature: Adding a significant new functionality to the game. This can be small in code-size, but is meant for the bigger things from a player perspective.
* Add: Similar to Feature, but for small functionalities.

> * Change: Changing existing behaviour to an extend the player needs to know about it.

Reviewer: ```suggestion
* Change: Changing existing behaviour to an extent the player needs to know about it.
```

* Fix: Fixing an issue with the game (as seen by the player).
* Remove: Completely removing a functionality.
* Revert: Reverting an earlier Feature / Add / Change / Fix / Remove.

> * Doc: Update to documentation.

Reviewer: Is this developer-facing docs such as source code comments, or user-facing docs? If the former, shouldn't it be under "Developer-facing changes"?

Author: It is under the player-facing part, so player-face :) The developer-facing is under `Codefix`, mostly :)

* Update: Translator commits.

> When a commit influence the moddable content, the above is also used, followed by a `[NewGRF]` or `[Script]` section.

Reviewer: ```suggestion
When a commit influences moddable content, the above is also used, followed by a `[NewGRF]` or `[Script]` tag.
```

Author: we call it `section` in the line a bit higher.

Reviewer: Hm, might be worth changing to tag throughout. Something like `[NewGRF]` is not a 'section' from the POV of a commit message, though it will eventually become a 'section' in the changelog.

Author: "sections" read fine to me, throughout this part. So not sure if renaming them is actually useful. But if others agree, I also have no problems changing it.

This means the commit is aimed (and worded) towards the NewGRF / Script authors, rather than players.

Developer-facing changes:
* Codechange: Changes to the code the player is not going to notice. Refactors, modernization, etc.
* Cleanup: Similar to Codechange, but when it is more about removing old code, rather than an actual change.
* Codefix: Fixing problems in earlier commits that the player is not actually going to notice. Wrong comments, missing files, CI changes.

> If you commit a `Fix` for an [issue](https://github.com/OpenTTD/OpenTTD/issues), add the corresponding issue number in the form of #NNNN.

Reviewer: ```suggestion
If you commit a `Fix` for an [issue](https://github.com/OpenTTD/OpenTTD/issues), add the corresponding issue number in the form of #NNNNN.
```
Most issues will have five digits for the foreseeable future :)

> In the case of `Fix`es, if you know what commit hash the bug was introduced (eg regression), please mention that hash (the first 7 characters) as well just after the keyword (or, if present, after the issue number).

Reviewer: ```suggestion
In the case of `Fix`es, if you know the hash of the commit in which the bug was introduced (eg regression), please mention that hash (the first 7 characters) as well just after the keyword (or, if present, after the issue number).
```

> Finding the trouble-causing hash is highly encouraged as it makes backporting / branching / releases that much easier.

Reviewer: ```suggestion
Finding the trouble-causing commit is highly encouraged as it makes backporting / branching / releases that much easier.
```

Do not mention two keywords; if two apply, pick one that best represents the commit (for example, "Fix #123" is mostly always better than "Revert", even if both are true).

The `<Details>` part starts with a capital and does not end with a dot.
Try to be descriptive to what the player will notice, not to what is actually being changed in the code.
See `changelog.txt` for inspiration.

To further structure the changelog, you can add sections. Example are:
* "Network" for network specific changes.
* "NewGRF" for NewGRF additions.
* "Script" for AI / GS additions.

> * "YAPP", "NPF", for changes in these features.

Reviewer: ```suggestion
* "YAPF", "NPF", for changes in these features.
```

* "MacOS", "Linux", "Windows", for OS-specific changes.
* "CI", "CMake", for changes to the (build) infrastructure.

Further explanations, more details, etc. don't go into the first line. Use a new line for those.

Complete examples:
* `Fix: [YAPF] Infinite loop in pathfinder`
* `Fix #5926: [YAPF] Infinite loop in pathfinder`
* `Codefix 80dffae: Warning about unsigned unary minus`

> * `Fix #6673, 99bb3a9: Store the map variety setting in the savegame`

Reviewer 2: Should it be "commit, issue", or "issue, commit"? The example is "issue, commit", but text earlier says "revision [...] just after the keyword".

Author: Added some words for the example to be correct :) At least, I think that is the best, as the ticket is the more important piece of information (to the player).

Ofc let me know if you disagree :)

Reviewer 2: I'm fine with either way, as long as the words are consistent :)
